compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -127,70 +152,44 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
             ast::ItemKind::MacroDef(ref macro_def) if macro_def.macro_rules => {
-                let is_macro_export =
-                    item.attrs.iter().any(|attr| attr.has_name(sym::macro_export));
-                if is_macro_export { Some(AccessLevel::Public) } else { None }
-            }
-
-            // Foreign modules inherit level from parents.
-            ast::ItemKind::ForeignMod(..) => self.prev_level,
-
-            // Other `pub` items inherit levels from parents.
-            ast::ItemKind::ExternCrate(..)
-            | ast::ItemKind::Use(..)
-            | ast::ItemKind::Static(..)
-            | ast::ItemKind::Const(..)
-            | ast::ItemKind::Fn(..)
-            | ast::ItemKind::Mod(..)
-            | ast::ItemKind::GlobalAsm(..)
-            | ast::ItemKind::TyAlias(..)
-            | ast::ItemKind::Enum(..)
-            | ast::ItemKind::Struct(..)
-            | ast::ItemKind::Union(..)
-            | ast::ItemKind::Trait(..)
-            | ast::ItemKind::TraitAlias(..)
-            | ast::ItemKind::MacroDef(..) => {
-                if item.vis.kind.is_pub() {
-                    self.prev_level
+                if item.attrs.iter().any(|attr| attr.has_name(sym::macro_export)) {
+                    self.set_access_level(item.id, Some(AccessLevel::Public));
                 } else {
-                    None
+                    self.set_access_level(item.id, None);
                 }
             }
 
-            // Should be unreachable at this stage
-            ast::ItemKind::MacCall(..) => panic!(
-                "ast::ItemKind::MacCall encountered, this should not anymore appear at this stage"
-            ),
-        };
-
-        let access_level = self.set_access_level(item.id, inherited_item_level);
-
-        // Set access level of nested items.
-        // If it's a mod, also make the visitor walk all of its items
-        match item.kind {
             ast::ItemKind::Mod(..) => {
+                access_level = self.set_access_level(item.id, access_level);
+                let orig_level = std::mem::replace(&mut self.prev_level, access_level);
                 if access_level.is_some() {
                     self.set_exports_access_level(self.r.local_def_id(item.id));
                 }
-
-                let orig_level = std::mem::replace(&mut self.prev_level, access_level);
                 visit::walk_item(self, item);
                 self.prev_level = orig_level;
             }
 
-            ast::ItemKind::ForeignMod(ForeignMod { ref items, .. }) => {
-                for nested in items {
-                    if nested.vis.kind.is_pub() {
-                        self.set_access_level(nested.id, access_level);
-                    }
-                }
-            }
             ast::ItemKind::Enum(EnumDef { ref variants }, _) => {
                 for variant in variants {
                     let variant_level = self.set_access_level(variant.id, access_level);
```

Comment:
Constructors in variants and structures don't need to be visited if they are processed in `fn set_exports_access_level` (which at this point can be renamed to `fn set_bindings_access_level`).

Reply:
In the end we should only visit fields, `ForeignMod`s, and `MacroDef`s in `fn visit_item`, everything else will be done using `NameBinding` visiting.